### PR TITLE
Add CombinedCertificateValidationContext.

### PR DESCRIPTION
**Pull request by JimmyCYJ**
Signed-off-by: JimmyCYJ <jimmychen.0102@gmail.com>

Description: Implement a new certificate validation context type CombinedCertificateValidationContext, which has a default CertificateValidationContextoption and SDS config. This default CertificateValidationContext will be merged with dynamic CertificateValidationContext into a new secret to serve. This is option 4 in https://docs.google.com/document/d/12gdjGN5m3v4vxUnDAglCP6pyyMoeuVGAGo7D_jc27jw/edit?usp=sharing
*Risk Level*: Low
*Testing*:
*Docs Changes*: NONE
*Release Notes*: NONE
*Fixes*: #4763

### CODE REVIEW

File: api/envoy/api/v2/auth/cert.proto [hunk 1]
```diff
@@ -238,12 +238,27 @@ message CommonTlsContext {
...
     CertificateValidationContext validation_context = 3;
 
     // Config for fetching validation context via SDS API.
     SdsSecretConfig validation_context_sds_secret_config = 7;
+
+    // Combined certificate validation context holds a default CertificateValidationContext
+    // and SDS config. When SDS server returns dynamic CertificateValidationContext, both dynamic
+    // and default CertificateValidationContext are merged into a new CertificateValidationContext
+    // for validation. This merge is done by Message::MergeFrom(), so singular fields will be
+    // overwritten, and repeated fields will be concatenated.
```
Comment (qiwzhang) [inline api/envoy/api/v2/auth/cert.proto]: not clear who override who?  sds override the "default"?

File: source/common/ssl/context_config_impl.cc [hunk 2]
```diff
@@ -45,32 +47,57 @@ Secret::TlsCertificateConfigProviderSharedPtr getTlsCertificateConfigProvider(
...
-          sds_secret_config.sds_config(), sds_secret_config.name(), factory_context);
-    }
+    envoy::api::v2::auth::Secret default_secret;
+    return getCertificateValidationContextConfigProviderFromSds(sds_secret_config, default_secret);
+  }
+  case envoy::api::v2::auth::CommonTlsContext::ValidationContextTypeCase::
+      kCombinedValidationContext: {
+    envoy::api::v2::auth::Secret default_secret;
+    default_secret.mutable_validation_context()->CopyFrom(
+        config.combined_validation_context().default_validation_context());
```
Comment (qiwzhang) [inline source/common/ssl/context_config_impl.cc]: Not need to make a copy. Just do

const auto& default_secret = config.combined_validation_context().default_validation_context();

BTW,  setDefaultSecret() not need to be at  base class.  It can be specific to CertificateValidationContextProvider

File: api/envoy/api/v2/auth/cert.proto (hunk 1 above)
Comment (JimmyCYJ) [reply api/envoy/api/v2/auth/cert.proto]: Comment is updated.

File: source/common/ssl/context_config_impl.cc (hunk 2 above)
Comment (JimmyCYJ) [reply source/common/ssl/context_config_impl.cc]: Done. Thanks.

**Comment by qiwzhang** [comment]
LGTM

File: source/common/secret/secret_manager_impl.cc [hunk 3]
```diff
@@ -34,6 +34,9 @@ void SecretManagerImpl::addStaticSecret(const envoy::api::v2::auth::Secret& secr
     }
     break;
   }
+  // Static secret of type
```
Comment (lizan) [inline source/common/secret/secret_manager_impl.cc]: no need this comment, CombinedValidationContext is not in oneof in Secret

File: source/common/ssl/context_config_impl.cc [hunk 4]
```diff
@@ -45,32 +46,60 @@ Secret::TlsCertificateConfigProviderSharedPtr getTlsCertificateConfigProvider(
...
+    if (!secret_provider) {
+      throw EnvoyException(fmt::format("Unknown static certificate validation context: {}",
+                                       sds_secret_config.name()));
+    }
+    return secret_provider;
+  } else {
+    auto secret_provider =
+        factory_context.secretManager().findOrCreateCertificateValidationContextProvider(
+            sds_secret_config.sds_config(), sds_secret_config.name(), factory_context);
+    if (default_cvc != nullptr) {
```
Comment (lizan) [inline source/common/ssl/context_config_impl.cc]: Isn't this semantically wrong? If secretManager() find existing CertificateValidationContextProvider then you're overriding the existing default CVC. If you have multiple cluster with same sds config name but different default_cvc, (I think this is common in istio use case, root cert from SDS are same, but secure naming is different) the last one wins and applies to all cluster.

File: source/common/secret/sds_api.h [hunk 5]
```diff
@@ -152,14 +152,27 @@ class CertificateValidationContextSdsApi : public SdsApi,
     return update_callback_manager_.add(callback);
   }
 
+  // Sets a default CertificateValidationContext. Once the default secret is set,
+  // it will be merged with dynamic CertificateValidationContext as new
+  // secret to provide.
+  void setDefaultSecret(const envoy::api::v2::auth::CertificateValidationContext& default_secret) {
+    default_cvc_.CopyFrom(default_secret);
```
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply source/common/secret/sds_api.h]: Done.

File: source/common/secret/secret_manager_impl.cc (hunk 3 above)
Comment (JimmyCYJ) [reply source/common/secret/secret_manager_impl.cc]: Removed. Thanks!

File: source/common/ssl/context_config_impl.cc (hunk 4 above)
Comment (JimmyCYJ) [reply source/common/ssl/context_config_impl.cc]: That is a good catch. Thanks, Lizan! I modify findOrCreateCertificateValidationContextProvider() to add default cvc as extra key. If we have multiple Clusters which use different default cvc and same SDS config, then secret manager creates different secret providers. Only clusters with the same cvc and SDS config share one provider.

File: include/envoy/secret/secret_manager.h [hunk 6]
```diff
@@ -70,11 +70,13 @@ class SecretManager {
    * @param config_name a name that uniquely refers to the SDS config source.
    * @param secret_provider_context context that provides components for creating and initializing
    * secret provider.
+   * @param extra_key that is needed to find a dynamic secret provider.
```
Comment (qiwzhang) [inline include/envoy/secret/secret_manager.h]: rename extra_key to extra_sharing_key.

extra_sharing_key is to control how secret providers can be shared. Usually they are shared if sds_config is the same. But for CombinedValidationContext case, they should not be share if their default_validation_context is not the same.  the extra_sharing_key will pass in default_validation_context

File: include/envoy/secret/secret_manager.h [hunk 7]
```diff
@@ -70,11 +70,13 @@ class SecretManager {
...
    * @return TlsCertificateConfigProviderSharedPtr the dynamic TLS secret provider.
    */
   virtual TlsCertificateConfigProviderSharedPtr findOrCreateTlsCertificateProvider(
       const envoy::api::v2::core::ConfigSource& config_source, const std::string& config_name,
-      Server::Configuration::TransportSocketFactoryContext& secret_provider_context) PURE;
+      Server::Configuration::TransportSocketFactoryContext& secret_provider_context,
+      const std::string& extra_key) PURE;
```
Comment (lizan) [inline include/envoy/secret/secret_manager.h]: What is the use case for extra_key for TlsCertificateProvider? it isn't used anywhere?

File: api/envoy/api/v2/auth/cert.proto [hunk 8]
```diff
@@ -238,12 +238,29 @@ message CommonTlsContext {
   // Configs for fetching TLS certificates via SDS API.
   repeated SdsSecretConfig tls_certificate_sds_secret_configs = 6;
 
+  message CombinedCertificateValidationContext {
+    // How to validate peer certificates.
+    CertificateValidationContext default_validation_context = 1;
```
Comment (lizan) [inline api/envoy/api/v2/auth/cert.proto]: mark this required

File: api/envoy/api/v2/auth/cert.proto [hunk 9]
```diff
@@ -238,12 +238,29 @@ message CommonTlsContext {
...
+
+    // Config for fetching validation context via SDS API.
+    SdsSecretConfig validation_context_sds_secret_config = 2;
```
Comment (lizan) [inline api/envoy/api/v2/auth/cert.proto]: mark this required too, if one of those are missing it shouldn't be combined cvc

File: api/envoy/api/v2/auth/cert.proto [hunk 10]
```diff
@@ -238,12 +238,29 @@ message CommonTlsContext {
...
     // Config for fetching validation context via SDS API.
     SdsSecretConfig validation_context_sds_secret_config = 7;
+
+    // Combined certificate validation context holds a default CertificateValidationContext
+    // and SDS config. When SDS server returns dynamic CertificateValidationContext, both dynamic
+    // and default CertificateValidationContext are merged into a new CertificateValidationContext
+    // for validation. This merge is done by Message::MergeFrom(), so dynamic
+    // CertificateValidationContext overwrites singular fields in default
+    // CertificateValidationContext, and concatenates repeated fields to default
+    // CertificateValidationContext.
```
Comment (lizan) [inline api/envoy/api/v2/auth/cert.proto]: I think we need some extra check during init for this (and document here) that until both default and overriding cvc are received from control plane, the cluster / listener will reject connections. Empty trusted ca means no validation, so it is important to make sure both are received.

File: source/common/ssl/context_config_impl.cc (hunk 4 above)
Comment (lizan) [reply source/common/ssl/context_config_impl.cc]: OK it makes this correct in Envoy side, but in this case you're flooding SDS server with O(# of unique clusters + # of unique listeners) request with same fetch trusted CA CVC request. Do you want to optimize that a bit?

File: include/envoy/secret/secret_manager.h (hunk 6 above)
Comment (JimmyCYJ) [reply include/envoy/secret/secret_manager.h]: Done.

File: include/envoy/secret/secret_manager.h (hunk 7 above)
Comment (JimmyCYJ) [reply include/envoy/secret/secret_manager.h]: It is removed from findOrCreateTlsCertificateProvider.

File: source/common/ssl/context_config_impl.cc (hunk 4 above)
Comment (JimmyCYJ) [reply source/common/ssl/context_config_impl.cc]: @lizan I have updated this PR with a change to provider::secret() interface. Now we keep secret config (e.g. TlsCertificateConfigPtr and CertificateValidationContextConfigPtr into ContextConfigImpl) Secret provider only returns secret() in protobuf message. The secret config will be created when SdsApi::update_callback_manager_ is invoked. Could you please take a look? Thanks!

File: api/envoy/api/v2/auth/cert.proto (hunk 10 above)
Comment (JimmyCYJ) [reply api/envoy/api/v2/auth/cert.proto]: I think with the latest change, we don't need extra check. ContextConfigImpl::isReady() tells us if secret config is ready. Only when received CVC and default CVC can merge into a valid CVC, the cluster / listener will use the secret to handle connections.

File: source/common/ssl/context_config_impl.cc (hunk 4 above)
Comment (JimmyCYJ) [reply source/common/ssl/context_config_impl.cc]: @lizan I will add / update tests when this change looks good to proceed.

File: include/envoy/secret/secret_manager.h [hunk 11]
```diff
@@ -70,6 +70,7 @@ class SecretManager {
    * @param config_name a name that uniquely refers to the SDS config source.
    * @param secret_provider_context context that provides components for creating and initializing
    * secret provider.
+   * @param extra_key that is needed to find a dynamic secret provider.
```
Comment (lizan) [inline include/envoy/secret/secret_manager.h]: you don't need this line?

File: source/common/ssl/context_config_impl.h [hunk 12]
```diff
@@ -26,12 +26,10 @@ class ContextConfigImpl : public virtual Ssl::ContextConfig {
   const std::string& cipherSuites() const override { return cipher_suites_; }
   const std::string& ecdhCurves() const override { return ecdh_curves_; }
   const TlsCertificateConfig* tlsCertificate() const override {
-    return tls_certficate_provider_ == nullptr ? nullptr : tls_certficate_provider_->secret();
+    return tls_certificate_config_ == nullptr ? nullptr : tls_certificate_config_.get();
```
Comment (lizan) [inline source/common/ssl/context_config_impl.h]: is this just `tls_certificate_config_.get()` now?

File: api/envoy/api/v2/auth/cert.proto [hunk 13]
```diff
@@ -238,12 +238,31 @@ message CommonTlsContext {
...
     CertificateValidationContext validation_context = 3;
 
     // Config for fetching validation context via SDS API.
     SdsSecretConfig validation_context_sds_secret_config = 7;
+
+    // Combined certificate validation context holds a default CertificateValidationContext
+    // and SDS config. When SDS server returns dynamic CertificateValidationContext, both dynamic
+    // and default CertificateValidationContext are merged into a new CertificateValidationContext
+    // for validation. This merge is done by Message::MergeFrom(), so dynamic
+    // CertificateValidationContext overwrites singular fields in default
```
Comment (htuch) [inline api/envoy/api/v2/auth/cert.proto]: What about `allow_expired_certificate`, which is non-wrapped; is that really overwrite or `||` logic?

File: source/common/ssl/context_config_impl.cc [hunk 14]
```diff
@@ -99,13 +136,66 @@ ContextConfigImpl::ContextConfigImpl(
...
+
+Ssl::CertificateValidationContextConfigPtr ContextConfigImpl::getCombinedValidationContextConfig(
+    const envoy::api::v2::auth::CertificateValidationContext& dynamic_cvc) {
+  envoy::api::v2::auth::CertificateValidationContext combined_cvc = *default_cvc_.get();
+  combined_cvc.MergeFrom(dynamic_cvc);
+  return std::make_unique<CertificateValidationContextConfigImpl>(combined_cvc);
+}
+
+void ContextConfigImpl::setSecretUpdateCallback(std::function<void()> callback) {
+  if (tls_certficate_provider_) {
```
Comment (htuch) [inline source/common/ssl/context_config_impl.cc]: Can I ask for a lot more comments on each of the basic blocks here and elsewhere? It's kind of hard to follow this unless you already have deep implementation knowledge.

File: source/common/ssl/context_config_impl.cc [hunk 15]
```diff
@@ -99,13 +136,66 @@ ContextConfigImpl::ContextConfigImpl(
...
+      max_protocol_version_(
+          tlsVersionFromProto(config.tls_params().tls_maximum_protocol_version(), TLS1_2_VERSION)) {
+  if (default_cvc_) {
+    cvc_validation_callback_handle_ =
+        dynamic_cast<Secret::CertificateValidationContextSdsApi*>(
+            certficate_validation_context_provider_.get())
+            ->addValidationCallback(
+                [this](const envoy::api::v2::auth::CertificateValidationContext& dynamic_cvc) {
+                  getCombinedValidationContextConfig(dynamic_cvc);
+                  validation_context_config_.reset();
```
Comment (qiwzhang) [inline source/common/ssl/context_config_impl.cc]: dont free validation_context_config_ here.  In case this config is bad (bad from other default config impl, not from this one),  we still want to use old config.

File: api/envoy/api/v2/auth/cert.proto (hunk 13 above)
Comment (JimmyCYJ) [reply api/envoy/api/v2/auth/cert.proto]: Yes, it is logical OR. I have verified this field in sds_api_test.cc. Thanks.

File: source/common/ssl/context_config_impl.cc (hunk 15 above)
Comment (JimmyCYJ) [reply source/common/ssl/context_config_impl.cc]: Good point. Removed. Thanks.

File: source/common/ssl/context_config_impl.cc (hunk 14 above)
Comment (JimmyCYJ) [reply source/common/ssl/context_config_impl.cc]: Add comments for this part. Thanks.

**Comment by JimmyCYJ** [comment]
@lizan @htuch This is ready for review. Please take a look. Thanks.

File: source/common/ssl/context_config_impl.cc [hunk 16]
```diff
@@ -18,14 +21,17 @@ namespace {
 
 Secret::TlsCertificateConfigProviderSharedPtr getTlsCertificateConfigProvider(
     const envoy::api::v2::auth::CommonTlsContext& config,
-    Server::Configuration::TransportSocketFactoryContext& factory_context) {
+    Server::Configuration::TransportSocketFactoryContext& factory_context,
+    Ssl::TlsCertificateConfigPtr* tls_config) {
```
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply source/common/ssl/context_config_impl.cc]: tls_config is not base config. tls_config holds a copy of TLS certificate config that is created from tls secret provider. 

default_cvc_ holds the default certificate validation context, and I have put a comment in context_config_impl.h introducing this new member.

I also simplify getCertificateValidationContextConfigProvider() and getTlsCertificateConfigProvider() by moving the creation of tls_certificate_config_ and validation_context_config_ to the body of ContextConfigImpl constructor.

File: source/common/ssl/context_config_impl.cc [hunk 17]
```diff
@@ -45,32 +52,62 @@ Secret::TlsCertificateConfigProviderSharedPtr getTlsCertificateConfigProvider(
   return nullptr;
 }
 
+Secret::CertificateValidationContextConfigProviderSharedPtr
+getProviderFromSds(Server::Configuration::TransportSocketFactoryContext& factory_context,
+                   const envoy::api::v2::auth::SdsSecretConfig& sds_secret_config,
+                   Ssl::CertificateValidationContextConfigPtr* cvc_config) {
+  if (!sds_secret_config.has_sds_config()) {
```
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply source/common/ssl/context_config_impl.cc]: I have inverted this loop and the loop in getTlsCertificateConfigProvider() accordingly.

File: source/common/ssl/context_config_impl.cc [hunk 18]
```diff
@@ -99,13 +136,80 @@ ContextConfigImpl::ContextConfigImpl(
...
+    // But the combined certificate validation context should pass validation. If
+    // validation of combined certificate validation context fails,
+    // getCombinedValidationContextConfig() throws exception, validation_context_config_ will not
+    // get updated.
+    cvc_validation_callback_handle_ =
+        dynamic_cast<Secret::CertificateValidationContextSdsApi*>(
+            certficate_validation_context_provider_.get())
+            ->addValidationCallback(
+                [this](const envoy::api::v2::auth::CertificateValidationContext& dynamic_cvc) {
+                  getCombinedValidationContextConfig(dynamic_cvc);
```
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply source/common/ssl/context_config_impl.cc]: Yes, the constructor of CertificateValidationContextConfigImpl will validate the combined certificate validation context. If validation fails, it will throw exception.

File: source/common/ssl/context_config_impl.h [hunk 19]
```diff
@@ -26,41 +26,26 @@ class ContextConfigImpl : public virtual Ssl::ContextConfig {
...
-    // Either tls_certficate_provider_ is nullptr or
-    // tls_certficate_provider_->secret() is NOT nullptr and
-    // either certficate_validation_context_provider_ is nullptr or
-    // certficate_validation_context_provider_->secret() is NOT nullptr.
-    return (!tls_certficate_provider_ || tls_certficate_provider_->secret() != nullptr) &&
-           (!certficate_validation_context_provider_ ||
-            certficate_validation_context_provider_->secret() != nullptr);
+    bool tls_is_ready = (tls_certficate_provider_ == nullptr || tls_certificate_config_ != nullptr);
+    bool combined_cvc_is_ready = (default_cvc_ == nullptr || validation_context_config_ != nullptr);
+    bool cvc_is_ready = (certficate_validation_context_provider_ == nullptr ||
```
[the remark this answers is not in the document]
Comment (JimmyCYJ) [reply source/common/ssl/context_config_impl.h]: Fixed. Thanks.

**Comment by JimmyCYJ** [comment]
@htuch This is ready to review. Please take a look. Thanks.
